### ISSUE TITLE
A timeline for displaying music scores

### ISSUE THREAD

**Issue by FelipeDefensor** [TimeLineAnnotator/desktop#219]
It is time to address the elephant in the room, which is displaying musical notation.

Conventional scores and TiLiA use space differently in a fundamental way: horizontal space is always to-scale in relation to time in TiLiA, while in conventional scores that relation is frequently relaxed in order to avoid overlaps or packing symbols too close together. There are no similar constraints in the use of vertical space, on the other hand.

With that in mind, I think that a good solution is to go with a mix between a piano roll and a conventional score, using a to-scale horizontal axis and a staff-based vertical axis. If I didn't believe a picture speaks a thousand words, I wouldn't have made TiLiA, so here's what I mean. This:

![example-score](https://github.com/user-attachments/assets/4499cd29-86bc-4c8d-b571-695ea862ba76)

Would be displayed in the score timeline as:

![Q8q03TO8hD1Xnaau](https://github.com/user-attachments/assets/5acc3c9d-d598-4a4e-a7ab-1da8a5c3262a)

Ideally, the data structures will be set up in a way that will also allow for a piano roll and conventional UIs to be easily implemented on top of them, later.

There are some things to be sorted out if we choose to go with this representations, for instance:

- How to display simultaneously seconds? Their rectangles would overlap.
- How to display accidentals when another note would obstruct them?
- Where to display symbols like clefs and time signatures? They would overlap with notes if displayed directly on top of the staff.

We definitely want to be able to import from conventional formats like MusicXML, which will require writing some kind of parser to our internal format. I have not yet researched about it, but I imagine there is already some available Python tooling. In the worst case, we can wrangle the format ourselves.

Another option is to go straight into conventional notation using tools like Verovio. That is something I am not sure I am ready to tackle, due to the issues that probably will arise with the differences in dealing with horizontal space.

**Comment by FelipeDefensor** [comment]
This was originally a feature that I intended to work on sometime in 2025, but that I decided to implement right now due to the needs of my PhD, of the Sinergia project and of the Anlysis-a-Thon taking place early 2025.

**Comment by FelipeDefensor** [comment]
@azfoo I don't know if you have given this feature some thought already. Please let me know you opinion.

**Comment by FelipeDefensor** [comment]
This issue #218 mentions some features that we want to have in place before releasing the score timeline.

**Comment by azfoo** [comment]
I'm not sure about displaying music with blocks - it isn't the most readable and there will be a (re)learning curve to reading scores in the app. Perhaps we could go with using the space created between beats to fit in all of the notes and accidentals attached to that beat somewhat equally spaced by their time value and use conventional notation. We could even float the clef, time signature and key signature on the left of the timeline, so that it is always there. 

Regarding parsing MusicXML, I've read up a little on the documentation and converting directly from that to TiLiA elements shouldn't be complicated enough to require an external library.

**Comment by FelipeDefensor** [comment]
> I'm not sure about displaying music with blocks - it isn't the most readable and there will be a (re)learning curve to reading scores in the app.

I agree those are tradeoffs to be considered.

> Perhaps we could go with using the space created between beats to fit in all of the notes and accidentals attached to that beat somewhat equally spaced by their time value and use conventional notation.

That would be great, but it only works if we are zoomed in quite a lot. The typical zoom levels I'm used to work would result in very ugly overlaps, I imagine. Can we get around that? I was thinking maybe a parallax could do the trick. We keep the score zoomed in so it does not overlap. but scroll it a faster rate than the rest of the timelines. Even then, we lose what might be a helpful zoomed out overview.

> We could even float the clef, time signature and key signature on the left of the timeline, so that it is always there.

That's a good idea, but we would still need to figure out a way of showing them aligned with the notes sometimes to signal clef/signature changes.

> Regarding parsing MusicXML, I've read up a little on the documentation and converting directly from that to TiLiA elements shouldn't be complicated enough to require an external library.

That's good to hear. If you dig around the `score-timeline` you can find the attributes I'm currently using and parse with that in mind. They are not complete nor final, of course, feel free to propose changes.

**Comment by azfoo** [comment]
> > Perhaps we could go with using the space created between beats to fit in all of the notes and accidentals attached to that beat somewhat equally spaced by their time value and use conventional notation.
> 
> That would be great, but it only works if we are zoomed in quite a lot. The typical zoom levels I'm used to work would result in very ugly overlaps, I imagine. Can we get around that? I was thinking maybe some parallax could do the trick. We keep the score zoomed in so it does not overlap. but scroll it a faster rate than the rest of the timelines. Even then, we lose what might be a helpful zoomed out overview.

Alternatively, we could have it as a QDialog that scrolls? I get zooming on a piano roll but on music notation.

**Comment by FelipeDefensor** [comment]
> Alternatively, we could have it as a QDialog that scrolls (something like a dockable PdfViewer but without page numbers)? I get zooming on a piano roll but not on music notation.

It could help with the display, but maybe interaction would become harder as it wouldn't share the `QGraphicsScene` with the rest of the timeline. Box selecting, for instance, wouldn't work.

The "blocks" are a mixture between conventional notation and piano rolls. One of the good things they would (supposedly) get from piano rolls is precisely remaning meaningful when zoomed out. I should have a prototype be the end of the week, so if you're able to spin up a minimal parser until then we can test with real scores to see how things look.

**Comment by azfoo** [comment]
I'll hopefully have a functioning parser in the next two days. 
Just to clarify because I couldn't figure out from the code: what do `step` and `octave` in `ComponentKinds.CLEF` mean?

**Comment by MarkGotham** [comment]
Hi both, and thanks for this!

Great to have this conversation. Here are some notes and "in principle" opinions.

Let's not assume that a score rolls in the same way as the other, continuous timelines. The current approach of "page turns" works well IMO. Plenty of room for improvement on that, but many have thought about this problem and settled on staying with something like the traditional static score "page" (broadly defined) and controlling page turns. A good example is performers using ipads and similar: there's never a rolling score; always fixed pages and often a device to allow the performed to chose when to turn.

How best to make that work in our case is open to design. One approach is to highlight (e.g., bounding box) the area of the other timelines that's currently being displayed on the score side.

To be clear, I'm not talking about static images in the sense of an imported PDF here. I definitely support the move to semantic score reading and rendering. Have a look at notable precedents for part/all of this equation, e.g., Dezrann, Vexflow, and Verovio. 

Speaking of Verovio, conversion of our scores via MEI is potentially a big project, but worth considering for this among other reasons.

Hope that's clear and helps a bit. Happy to chat sometime.

**Comment by FelipeDefensor** [comment]
> I'll hopefully have a functioning parser in the next two days. Just to clarify because I couldn't figure out from the code: what do `step` and `octave` in `ComponentKinds.CLEF` mean?

Step is zero-based numeral correspoding to the "note name": C is 0, D is 1, ..., B is 6. `ComponentKinds.Clef` is the component kind for clefs. As those will need to be positioned and will also be involved in caclcuting note positions, they will be a component like everything else.

**Comment by FelipeDefensor** [comment]
> Could we have conventional notation in a dock widget but in one line that "scrolls" (more like a quick horizontal shift) when the next measure is about to be out of view? The timeline version of this could be a zoomed out image of the same, that somewhat aligns with the beat timeline, with a bounding box indicating the position that the dock widget is showing. The scroll position of the timeline and dock widget can be linked.

I'm not sure I got that completely, but I think it might work, yes.

> That way, we don't actually need to handle making components (which will only make the total file size bigger) and instead render an SVG (maybe hosted online) with a csv on image x-position (start/end) to measure number.

Components are needed to make annotating score elements possible. I think that adding comments and coloring notes, at least, is a feature that we should have. Later, we could go towards having more complex graphical annotation features like circling or bracketing groups of notes, which are features that are very useful for analysis and but that conventional score editors support only poorly.

That being said, maybe we could create components only for the score elements that are annotated.

**Comment by FelipeDefensor** [comment]
> Let's not assume that a score rolls in the same way as the other, continuous timelines. The current approach of "page turns" works well IMO. Plenty of room for improvement on that, but many have thought about this problem and settled on staying with something like the traditional static score "page" (broadly defined) and controlling page turns. A good example is performers using ipads and similar: there's never a rolling score; always fixed pages and often a device to allow the performed to chose when to turn.

I really think we should have some timeline-like visualization, as that will make following the score much easier, and also connecting the score symbols to the rest of the annotations in TiLiA. I don't think performing a piece is a relevant use case for TiLiA. That said, I don't see why we couldn't have both, and probably the "vertical" score with page breaks will be easier to implement right now.

What about we keep moving to have a piano-roll/score mix like I suggested, one that does not display all the score symbols (as all those would significant more time to implement) but also have a conventional score based view using the bounding box like you suggested by leveraging Verovio or something similar?

**Comment by FelipeDefensor** [comment]
@azfoo I have just pushed the prototype.

**Comment by azfoo** [comment]
@FelipeDefensor mostly done with the parser - it needs a few more bits to fully integrate (see commit)

**Comment by azfoo** [comment]
@FelipeDefensor parser update: done
one more thing that would make things easier is to have a staff id on score components, so that multistaff music can be displayed.

**Comment by FelipeDefensor** [comment]
I think we should explore the conventional notation stuff for a bit before deciding what to prioritize next. I had no luck getting a Verovio SVG working in PyQt. Here's the script I'm trying to run:

```
import sys

import verovio
from PyQt6.QtGui import QPainter
from PyQt6.QtSvgWidgets import QSvgWidget
from PyQt6.QtWidgets import QApplication, QLabel

tk = verovio.toolkit()
tk.loadFile('example.mei')
svg_string = tk.renderToSVG()

svg_bytes = bytearray(svg_string, encoding='utf-8')

app = QApplication(sys.argv)
svgWidget = QSvgWidget()
svgWidget.load('example.svg')
svgWidget.setGeometry(100, 100, 300, 300)
svgWidget.show()
label = QLabel('dhsjkadhsjak')
label.show()
app.exec()
```

I'm pretty sure this is supposed to work. Could you try it on your end?

**Comment by azfoo** [comment]
update: exporting a file as svg from an engraver shows up fine so maybe it isn't PyQt but verovio

**Comment by FelipeDefensor** [comment]
> update: exporting a file as svg from an engraver shows up fine so maybe it isn't PyQt but verovio

Maybe they use different SVG specs?

**Comment by azfoo** [comment]
The most likely cause that I can find is that verovio saves fonts as symbols of size 0,0 and references that to print everything else. Qt doesn't like that size and throws it away so then the subsequent calls to reference that symbol doesn't render anything.

**Comment by FelipeDefensor** [comment]
> @FelipeDefensor parser update: done

I took a look at it and it seems to work well, except the first measure seems to be skipped when importing.

![image](https://github.com/user-attachments/assets/fce86958-28aa-4e6a-a6b7-4d26dfbc34bb)

Here are the .tla and the .mxl to reproduce.

[lvb-16bars.zip](https://github.com/user-attachments/files/17619242/lvb-16bars.zip)

**Comment by FelipeDefensor** [comment]
> @azfoo I took a look at it and it seems to work well, except the first measure seems to be skipped when importing.

I removed the anacrusis and now eveything works perfectly. Here's a demo:

https://github.com/user-attachments/assets/94f310f5-8f6a-4ca7-9395-539016383fb3

I believe this kind of visualization (or at least piano roll) is worth our time, while not excluding the obvious value of also having conventional notation.

**Comment by azfoo** [comment]
@FelipeDefensor I'm about to merge the vexflow viewer into the app. Should this be done as an alternate view to the blocks (on the same timeline, with the option to view/hide) or a different timeline?

**Comment by FelipeDefensor** [comment]
I like the alternate view idea. Anything against that? Ideally we would separate now and merge later, but there's too much boilerplate to creating a new timeline kind.

**Comment by azfoo** [comment]
@FelipeDefensor: I'm mostly done with the svg viewer.
I got the animation to work, but couldn't sort out the non-floatable docking. It crashes with movable and closable set with no debug message. Setting only closable works, but that means it cannot be re-docked into a different position in the main window.

**Comment by azfoo** [comment]
If you have time to spare, could you compare the current head of score timeline and [smooth-timer](https://github.com/azfoo/TiLiA/tree/smooth-timer)? The difference being smooth-timer calls `Post.PLAYER_CURRENT_TIME_CHANGED` more regularly so everything that relies on time change also gets updated more regularly (and therefore, hopefully, more smoothly without having to apply timers on the individual components).

**Comment by FelipeDefensor** [comment]
> If you have time to spare, could you compare the current head of mxl-parser and [smooth-timer](https://github.com/azfoo/TiLiA/tree/smooth-timer)? The difference being smooth-timer calls `Post.PLAYER_CURRENT_TIME_CHANGED` more regularly so everything that relies on time change also gets updated more regularly (and therefore, hopefully, more smoothly without having to apply timers on the individual components).

If you can wait, I can do that tomorrow. Is there enough time for the other listeners to process their callbacks even when we fire that event more frequently?

**Comment by azfoo** [comment]
Yes, I can wait.
At the moment, it seems things are processed alright. Lowering the frequency of updates is also possible if necessary.

**Comment by FelipeDefensor** [comment]
Things look good when working with small files, but they get unresponsive with large ones (LvB proof-of-concept, for instance). Also, I found it very hard to follow the continuously scrolling score. I think we should either go back to the page flipping, or ensure that the current score position is always in the center of the view and indicate that with a transparent band (like this https://www.youtube.com/watch?v=TUatfMtpuiQ&list=PL174497556325A2E9). I think the latter might be too hard to implement right now.

I also spotted some bugs:

- Measure trackers desyncs as music progresses.

https://github.com/user-attachments/assets/ca023c8a-cd5f-4765-877f-c4c45cc29c58

- Measure tracker only seeks to playback position if we are playing (is this intentional?)

- Error saving file (this is probably my fault):

```
Traceback (most recent call last):
  File "C:\Users\Felipe\dev\tilia-desktop\tilia\ui\actions.py", line 137, in callback
    post(request)
  File "C:\Users\Felipe\dev\tilia-desktop\tilia\requests\post.py", line 218, in post
    callback(*args, **kwargs)
  File "C:\Users\Felipe\dev\tilia-desktop\tilia\file\file_manager.py", line 110, in on_save_as_request
    self.save(get(Get.APP_STATE), path)
  File "C:\Users\Felipe\dev\tilia-desktop\tilia\file\file_manager.py", line 178, in save
    write_tilia_file_to_disk(TiliaFile(**data), str(path))
  File "C:\Users\Felipe\dev\tilia-desktop\tilia\file\common.py", line 28, in write_tilia_file_to_disk
    with open(path, "w", encoding="utf-8") as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 22] Invalid argument: 'TiLiA files (*.tla)'

```

**Comment by azfoo** [comment]
> Things look good when working with small files, but they get unresponsive with large ones (LvB proof-of-concept, for instance). 
Is this with the mxl-parser branch or smooth-timer? There is smoothing applied on both branches but in slightly different ways.

> Also, I found it very hard to follow the continuously scrolling score. I think we should either go back to the page flipping, or ensure that the current score position is always in the center of the view and indicate that with a transparent band.
I did have a version with the band but its location was highly inaccurate. It also started making things lag quite quickly since that involved reloading the SVG. 
I'm gonna take a look at moving from svg widget to qGraphicsSvgItem to see if that makes things run more efficiently. 

> * Measure trackers desyncs as music progresses.
Will try to fix the maths

> * Measure tracker only seeks to playback position if we are playing (is this intentional?)
Currently, yes. Also when you drag the orange box in the main window. When else should it seek? 

> * Error saving file (this is probably my fault):
Fixed in #236

**Comment by FelipeDefensor** [comment]
> Is this with the mxl-parser branch or smooth-timer? There is smoothing applied on both branches but in slightly different ways.

That was the `mxl-parser`. I will try the `smooth-timer`, sorry.

> I did have a version with the band but its location was highly inaccurate. It also started making things lag quite quickly since that involved reloading the SVG.

Do we currently have to reload the SVG on every change?

> Currently, yes. Also when you drag the orange box in the main window. When else should it seek?

I think pretty much whenever the player itself seeks.

**Comment by azfoo** [comment]
> > I did have a version with the band but its location was highly inaccurate. It also started making things lag quite quickly since that involved reloading the SVG.
> 
> Do we currently have to reload the entire SVG on every change to the drawn SVG elements?

Yes - it was alright when dealing with small edits, but not for redrawing the playback line with every seek made. Will update on this again when I'm done with graphicsSvgItem. 

> > Currently, yes. Also when you drag the orange box in the main window. When else should it seek?
> 
> I think pretty much whenever the player itself seeks.

Will do.

**Comment by FelipeDefensor** [comment]
@azfoo, maybe you already know this but there are some test failures on the `mxl-parser` branch. A quick look tells me they might be related to the `Get` requests for the current time.

**Comment by FelipeDefensor** [comment]
@azfoo I got the docking on the SVGViewer to be as we wanted. The user can now switch docking areas but the widget is not floatable. The initial height of the timeline and svg viewer is messed, something to solve later.

I had to rebase some changes from `dev` (which implemented that docking behavior to the inspector, for instance) so I had to force push to `mxl-parser`.

**Comment by FelipeDefensor** [comment]
We are in the final stretch! Here are some bugs we need to deal with:

- [X] Can't undo/redo import MusicXML (F, 04-12)
- [ ] Some of the highest notes are still getting cropped
- [ ] Ctrl+Z not working with 'Ambiguous shortcut overload'
- [ ] Measure tracker is not repositioned when zooming
- [ ] Measure tracker reappears when zooming even if it was hidden by user
- [ ] Measure tracker goes to random place when multimeasure rests are tracked
- [ ] Sometimes VexFlox SVG is generated in more than one line

**Comment by FelipeDefensor** [comment]
@azfoo I fixed undo/redo and listed to-dos above.

**Comment by FelipeDefensor** [comment]
> Sometimes VexFlox SVG is generated in more than one line

![image](https://github.com/user-attachments/assets/d2765fe8-1bdf-46ff-9984-291573a798a8)

@azfoo are we still using the browser or did we switch to the SVG widget?

**Comment by FelipeDefensor** [comment]
> Failing tests (due to no reply to Get.MAIN_WINDOW)

@azfoo test in `timelines/score/test_score_timeline.py` are failing because the SvgViewer constructor tries to get the main window, but since this is a test only for the backend, there no `QtUI` available to reply.

We should move the SVGViewer to ScoreTimelineUI to keep backend and frontend decoupled. I have a vague memory of we already having talked about this, sorry if that's the case. I figure I leave this fix up to you as the code has already changed considerably on your end.

**Comment by azfoo** [comment]
@FelipeDefensor: mostly done. If we can rebase to the work done on measure fractions, we should be good to go. 
The only other thing would be to look at QThreads as an option if performance is still an issue

**Comment by FelipeDefensor** [comment]
> @FelipeDefensor: mostly done. If we can rebase to the work done on measure fractions, we should be good to go. The only other thing would be to look at QThreads as an option if performance is still an issue

Great! Can't we merge both to dev separately?

Performance is fine for anything below 15k elements, I would say. Up until 25k elements (around what we get loading a Hauptstimme score) it is still acceptable, with zooming taking around 0.2s (on my machine). Why don't you try it and see what you think? There are certainly optimizations to de done, but I think all the low hanging fruits are gone.

I tried using multiprocessing to parallelize the zooming computations but the overhead for spawning the processes killed the performance gains. Might be worth trying again, though.

**Comment by FelipeDefensor** [comment]
> Fix tlui for multi stave music

Care to elaborate on this one? Or go ahead and fix it yourself if you know what's going on.

**Comment by FelipeDefensor** [comment]
Pulled the latest changes and now I'm getting this error when loading any SVG:

Traceback (most recent call last):
  File "C:\Users\Felipe\dev\tilia-desktop\tilia\ui\windows\svg_viewer.py", line 61, in set_svg
    self.on_svg_loaded(svg)
  File "C:\Users\Felipe\dev\tilia-desktop\tilia\ui\windows\svg_viewer.py", line 111, in preprocess_svg
    self.timeline.set_data("svg_data", svg)
  File "C:\Users\Felipe\dev\tilia-desktop\tilia\timelines\base\timeline.py", line 127, in set_data
    setattr(self, attr, value)
  File "C:\Users\Felipe\dev\tilia-desktop\tilia\timelines\score\timeline.py", line 79, in svg_data
    self.svg_view.load_svg_data(svg_data)
  File "C:\Users\Felipe\dev\tilia-desktop\tilia\ui\windows\svg_viewer.py", line 207, in load_svg_data
    for item in self.scene.items():
                ^^^^^^^^^^^^^^^^^^
RuntimeError: wrapped C/C++ object of type QGraphicsScene has been deleted

**Comment by azfoo** [comment]
> > Fix tlui for multi stave music
> 
> Care to elaborate on this one? Or go ahead and fix it yourself if you know what's going on.

Any music that has multiple staves (like double-lined piano music) crashes:
```Traceback (most recent call last):
  File "D:\TiLiA\TiLiA\tilia\ui\actions.py", line 144, in callback
    post(request)
  File "D:\TiLiA\TiLiA\tilia\requests\post.py", line 218, in post
    callback(*args, **kwargs)
  File "D:\TiLiA\TiLiA\tilia\ui\qtui.py", line 488, in on_import_from_csv
    post(Post.SCORE_TIMELINE_COMPONENTS_DESERIALIZED, timeline.id)
  File "D:\TiLiA\TiLiA\tilia\requests\post.py", line 218, in post
    callback(*args, **kwargs)
  File "D:\TiLiA\TiLiA\tilia\ui\timelines\score\timeline.py", line 396, in on_score_timeline_components_deserialized
    element.on_components_deserialized()
  File "D:\TiLiA\TiLiA\tilia\ui\timelines\score\element\staff.py", line 40, in on_components_deserialized
    self.update_position()
  File "D:\TiLiA\TiLiA\tilia\ui\timelines\score\element\staff.py", line 32, in update_position
    self.staff_lines.set_position(*self.line_args())
  File "D:\TiLiA\TiLiA\tilia\ui\timelines\score\element\staff.py", line 72, in set_position
    y = middle_y + y_offset - line_spacing / 2
        ~~~~~~~~~^~~~~~~~~~
TypeError: unsupported operand type(s) for +: 'NoneType' and 'float'```

**Comment by FelipeDefensor** [comment]
> > > Fix tlui for multi stave music
> > 
> > 
> > Care to elaborate on this one? Or go ahead and fix it yourself if you know what's going on.
> 
> Any music that has multiple staves (like double-lined piano music) crashes:
> 
> ```
>   File "D:\TiLiA\TiLiA\tilia\ui\actions.py", line 144, in callback
>     post(request)
>   File "D:\TiLiA\TiLiA\tilia\requests\post.py", line 218, in post
>     callback(*args, **kwargs)
>   File "D:\TiLiA\TiLiA\tilia\ui\qtui.py", line 488, in on_import_from_csv
>     post(Post.SCORE_TIMELINE_COMPONENTS_DESERIALIZED, timeline.id)
>   File "D:\TiLiA\TiLiA\tilia\requests\post.py", line 218, in post
>     callback(*args, **kwargs)
>   File "D:\TiLiA\TiLiA\tilia\ui\timelines\score\timeline.py", line 396, in on_score_timeline_components_deserialized
>     element.on_components_deserialized()
>   File "D:\TiLiA\TiLiA\tilia\ui\timelines\score\element\staff.py", line 40, in on_components_deserialized
>     self.update_position()
>   File "D:\TiLiA\TiLiA\tilia\ui\timelines\score\element\staff.py", line 32, in update_position
>     self.staff_lines.set_position(*self.line_args())
>   File "D:\TiLiA\TiLiA\tilia\ui\timelines\score\element\staff.py", line 72, in set_position
>     y = middle_y + y_offset - line_spacing / 2
>         ~~~~~~~~~^~~~~~~~~~
> TypeError: unsupported operand type(s) for +: 'NoneType' and 'float'```
> ```

Will look into it.

**Comment by FelipeDefensor** [comment]
@azfoo I can't reproduce the crash with multiple staves. I'm at the `xml-parser` branch. These are the two files that load just fine for me.
[multi-staff-works.zip](https://github.com/user-attachments/files/18129585/multi-staff-works.zip)

**Comment by FelipeDefensor** [comment]
> @azfoo I can't reproduce the crash with multiple staves. I'm at the `xml-parser` branch. These are the two files that load just fine for me. [multi-staff-works.zip](https://github.com/user-attachments/files/18129585/multi-staff-works.zip)

Also, here's a tla ready to import them: 
[score_test.zip](https://github.com/user-attachments/files/18129593/score_test.zip)
port them:

**Comment by FelipeDefensor** [comment]
> > @azfoo, I pulled the latest changes and now I'm getting this error when loading any SVG:
> > ```
> > Traceback (most recent call last):
> >   File "C:\Users\Felipe\dev\tilia-desktop\tilia\ui\windows\svg_viewer.py", line 61, in set_svg
> >     self.on_svg_loaded(svg)
> >   File "C:\Users\Felipe\dev\tilia-desktop\tilia\ui\windows\svg_viewer.py", line 111, in preprocess_svg
> >     self.timeline.set_data("svg_data", svg)
> >   File "C:\Users\Felipe\dev\tilia-desktop\tilia\timelines\base\timeline.py", line 127, in set_data
> >     setattr(self, attr, value)
> >   File "C:\Users\Felipe\dev\tilia-desktop\tilia\timelines\score\timeline.py", line 79, in svg_data
> >     self.svg_view.load_svg_data(svg_data)
> >   File "C:\Users\Felipe\dev\tilia-desktop\tilia\ui\windows\svg_viewer.py", line 207, in load_svg_data
> >     for item in self.scene.items():
> >                 ^^^^^^^^^^^^^^^^^^
> > RuntimeError: wrapped C/C++ object of type QGraphicsScene has been deleted
> > ```
> 
> this is also unfortunately related to deserializing. will try and see if I can fix it

Ok. If you're not able to fix this today, let me know, so I can try to find a workaround to keepo working on the other bugs. Will get things ready for deploying, in the meantime.

**Comment by azfoo** [comment]
> > > @azfoo, I pulled the latest changes and now I'm getting this error when loading any SVG:
> > > ```
> > > Traceback (most recent call last):
> > >   File "C:\Users\Felipe\dev\tilia-desktop\tilia\ui\windows\svg_viewer.py", line 61, in set_svg
> > >     self.on_svg_loaded(svg)
> > >   File "C:\Users\Felipe\dev\tilia-desktop\tilia\ui\windows\svg_viewer.py", line 111, in preprocess_svg
> > >     self.timeline.set_data("svg_data", svg)
> > >   File "C:\Users\Felipe\dev\tilia-desktop\tilia\timelines\base\timeline.py", line 127, in set_data
> > >     setattr(self, attr, value)
> > >   File "C:\Users\Felipe\dev\tilia-desktop\tilia\timelines\score\timeline.py", line 79, in svg_data
> > >     self.svg_view.load_svg_data(svg_data)
> > >   File "C:\Users\Felipe\dev\tilia-desktop\tilia\ui\windows\svg_viewer.py", line 207, in load_svg_data
> > >     for item in self.scene.items():
> > >                 ^^^^^^^^^^^^^^^^^^
> > > RuntimeError: wrapped C/C++ object of type QGraphicsScene has been deleted
> > > ```
> > 
> > 
> > this is also unfortunately related to deserializing. will try and see if I can fix it
> 
> Ok. If you're not able to fix this today, let me know, so I can try to find a workaround to keepo working on the other bugs. Will get things ready for deploying, in the meantime.

done!

**Comment by azfoo** [comment]
> > @azfoo I can't reproduce the crash with multiple staves. I'm at the `xml-parser` branch. These are the two files that load just fine for me. [multi-staff-works.zip](https://github.com/user-attachments/files/18129585/multi-staff-works.zip)
> 
> Also, here's a tla ready to import them: [score_test.zip](https://github.com/user-attachments/files/18129593/score_test.zip)

here's a non working file
[multistaff not work.zip](https://github.com/user-attachments/files/18131649/multistaff.not.work.zip)

**Comment by FelipeDefensor** [comment]
> > > @azfoo I can't reproduce the crash with multiple staves. I'm at the `xml-parser` branch. These are the two files that load just fine for me. [multi-staff-works.zip](https://github.com/user-attachments/files/18129585/multi-staff-works.zip)
> > 
> > 
> > Also, here's a tla ready to import them: [score_test.zip](https://github.com/user-attachments/files/18129593/score_test.zip)
> 
> here's a non working file [multistaff not work.zip](https://github.com/user-attachments/files/18131649/multistaff.not.work.zip)

The problem here seems to be that the one of the staves is empty. Will fix.

**Comment by FelipeDefensor** [comment]
> QtCriticalMsg when importing Mozart sonata

![image](https://github.com/user-attachments/assets/34f41752-cfed-4b7b-b604-4e5919368069)

@azfoo this seems to be a VexFlow thing. Here's the file to reproduce.

[k279.zip](https://github.com/user-attachments/files/18135713/k279.zip)

**Comment by FelipeDefensor** [comment]
I boldly merged mxl-parser to dev, but it still needs polishing before release.

**Comment by FelipeDefensor** [comment]
The performance of the scrolling is fine for smallish scores, but things get very choppy even e.g. with the LvB main melody score. I think improving this should be a priority after the release.

**Comment by azfoo** [comment]
I've got a couple of fixes to test and commit but will work on the performance after that

**Comment by FelipeDefensor** [comment]
The score timeline branches have been merged and we have a functional version, so I'm closing this. Let's open issues for the remaining things we need to work on, such as:

- Catch no Internet to download node modules before crash (YouTube/OSMD)
- Crash when importing MXL with empty staff
- QtCriticalMsg when importing Mozart sonata
- Reimporting MXL makes SVGViewer disappear
